webassembly/api: Allocate code data on C heap when running Python code.

dpgeorge opened the pull request:
Otherwise Emscripten allocates it on the Emscripten C stack, which will overflow for large amounts of code.

Fixes issue #14307.

dpgeorge committed:
webassembly/api: Allocate code data on C heap when running Python code.

Otherwise Emscripten allocates it on the Emscripten C stack, which will
overflow for large amounts of code.

Fixes issue #14307.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Find a compiled version of this PR here for testing: https://micropython.org/resources/wasm/micropython.mjs

WebReflection commented on ports/webassembly/api.js:
```diff
@@ -131,23 +131,31 @@ export async function loadMicroPython(options) {
             return proxy_convert_mp_to_js_obj_jsside_with_free(value);
         },
         runPythonAsync(code) {
+            const len = Module.lengthBytesUTF8(code);
```
if `len` always needs a `+ 1` (for null code `\x00` I suppose) why isn't this just `const len = Module.lengthBytesUTF8(code) + 1;` instead? Not a blocker, just a silly question!

ntoll replied:
Bravo

dpgeorge replied:
`Module.stringToUTF8` will always add a null byte at the end, so there must be enough room for it.

But, below in the call to `mp_js_do_exec_async`, we pass just `len`.

WebReflection commented:
I can confirm this fixes the presented use-case/issue we had 🙏

[index.txt](https://github.com/micropython/micropython/files/15013165/index.txt)

dpgeorge committed:
webassembly/api: Fix waiting for Emscripten module to be loaded.

In modularize mode, the `_createMicroPythonModule()` constructor must be
await'ed on, before `Module` is ready to use.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Also added a fix for initialising `Module` with the latest Emscripten.

WebReflection commented:
@dpgeorge do you mind publishing this latest version to *npm* so we can also bump it? Thanks!
